pkg/node-modules/nodeModuleCollector.go

```diff
@@ -81,33 +88,62 @@ func (t *Collector) readDependencyTree(dependency *Dependency) error {
 		if err != nil {
 			return err
 		}
-		t.AddDependencyMap(queue[i], dependency)
+		queue[i].parent = dependency
 	}
-
 	return nil
 }
 
-func (t *Collector) AddDependencyMap(childDependency *Dependency, parentDependency *Dependency) {
-	if t.DependencyMap == nil {
-		t.DependencyMap = make(map[string]*Dependency)
+func (t *Collector) writeToParentConflicDependency(d *Dependency) {
+	p := d.parent
+	last := d
+	for p != t.rootDependency {
+		if p.conflictDependency != nil {
+			if c, ok := p.conflictDependency[d.Name]; ok {
+				if c.Version == d.Version {
+					return
+				}
+				break
+			}
+		}
+		last = p
+		p = p.parent
+	}
+
+	if last.conflictDependency == nil {
+		last.conflictDependency = make(map[string]*Dependency)
 	}
+	last.conflictDependency[d.Name] = d
+}
 
-	name := childDependency.Name
-	if d, ok := t.DependencyMap[name]; ok {
-		if d.Version != childDependency.Version {
-			if parentDependency.conflictDependency == nil {
-				parentDependency.conflictDependency = make(map[string]*Dependency)
+func (t *Collector) processHoistDependencyMap() {
+	t.HoiestedDependencyMap = make(map[string]*Dependency)
+	for _, d := range t.allDependencies[1:] {
+		if e, ok := t.HoiestedDependencyMap[d.Name]; ok {
+			if e.Version != d.Version {
+				if d.parent == t.rootDependency {
+					t.HoiestedDependencyMap[d.Name] = d
+					t.writeToParentConflicDependency(e)
+				} else {
+					t.writeToParentConflicDependency(d)
+				}
 			}
-			parentDependency.conflictDependency[name] = childDependency
+		} else {
+			t.HoiestedDependencyMap[d.Name] = d
 		}
-	} else {
-		t.DependencyMap[name] = childDependency
+
 	}
 }
 
 func (t *Collector) processDependencies(list *map[string]string, nodeModuleDir string, isOptional bool, queue *[]*Dependency, queueIndex int) (int, error) {
 	unresolved := make([]string, 0)
-	for name := range *list {
+
+	names := make([]string, 0, len(*list))
+	for k := range *list {
+		names = append(names, k)
+	}
+	sort.Strings(names)
```

Comment:
The electron-builder pipeline(https://github.com/electron-userland/electron-builder/actions/runs/10856416053/job/30131018855) occasionally failed, and I finally found the root cause. The map in Golang doesn't maintain a fixed order, so sometimes it returns `[foo, ms]`, and other times it returns `[ms, foo]`. If the order is incorrect, it leads to different results. Now that we've fixed the order, it won't fail anymore.

![image](https://github.com/user-attachments/assets/29076c07-0f49-45b4-9668-7ddf8715810c)